File: apps/web/src/lib/panel.ts
export type PanelOnboardingStatus = "ready" | "pending" | "blocked" | "custom-needed";
export type PanelRunStatus = "pass" | "fail" | "timeout" | "blocked" | "not-run" | null;
export type PanelHealth = "healthy" | "warning" | "failing" | "unknown";

export type PanelProject = {
  project_key: string;
  project_name: string;
  repo_path: string;
  preset: string;
  onboarding_status: PanelOnboardingStatus;
  module_count: number;
  health: PanelHealth;
  source: string | null;
  triggered_by: string | null;
  status: PanelRunStatus;
  block_reason: string | null;
  started_at: string | null;
  finished_at: string | null;
  duration_sec: number | null;
};

export type PanelProjectsResponse = {
  generated_at: string;
  projects: PanelProject[];
};

export type PanelCheckerResult = {
  checker: "build" | "lint" | "typecheck" | "test" | "coverage";
  status: "pass" | "fail" | "timeout" | "skip";
  detail: string | null;
  duration_sec: number | null;
};

export type PanelModule = {
  module_name: string;
  stack: string | null;
  language: string | null;
  status: Exclude<PanelRunStatus, null>;
  coverage_pct: number | null;
  baseline_pct: number | null;
  coverage_gate_pct: number | null;
  coverage_delta_pct: number | null;
  coverage_parser: string | null;
  block_reason: string | null;
  checker_results: PanelCheckerResult[];
};

export type PanelLatestRun = {
  run_key: string;
  source: string | null;
  git_ref: string | null;
  git_sha: string | null;
  triggered_by: string | null;
  started_at: string | null;
  finished_at: string | null;
  duration_sec: number | null;
  strict_mode: boolean;
  status: Exclude<PanelRunStatus, null>;
  block_reason: string | null;
};

export type PanelProjectDetail = {
  project_key: string;
  project_name: string;
  repo_path: string;
  preset: string;
  onboarding_status: PanelOnboardingStatus;
  latest_run: PanelLatestRun;
  modules: PanelModule[];
};

export type PanelProjectDetailResponse = {
  generated_at: string;
  project: PanelProjectDetail;
};

type ApiOverviewProject = {
  project_key: string;
  project_name: string;
  preset: string;
  module_count: number;
  onboarding_status: PanelOnboardingStatus;
  health: PanelHealth;
  check_all: Exclude<PanelRunStatus, null>;
  last_run_at: string | null;
  block_reason: string | null;
};

type ApiProjectsResponse = {
  generated_at: string;
  projects: ApiOverviewProject[];
};

type ApiCheckerResult = PanelCheckerResult;

type ApiRunModule = {
  module_name: string;
  stack: string | null;
  language: string | null;
  status: Exclude<PanelRunStatus, null>;
  coverage_pct: number | null;
  baseline_pct: number | null;
  coverage_gate_pct: number | null;
  coverage_delta_pct: number | null;
  coverage_parser: string | null;
  block_reason: string | null;
  checker_results: ApiCheckerResult[];
};

type ApiLatestRunRef = {
  run_key: string;
  status: Exclude<PanelRunStatus, null>;
  block_reason: string | null;
  finished_at: string;
};

type ApiProjectDetail = {
  project_key: string;
  project_name: string;
  preset: string;
  onboarding_status: PanelOnboardingStatus;
  latest_run: ApiLatestRunRef | null;
  modules: ApiRunModule[];
};

type ApiProjectDetailResponse = {
  generated_at: string;
  project: ApiProjectDetail;
};

type ApiLatestRun = {
  run_key: string;
  source: string | null;
  git_ref: string | null;
  git_sha: string | null;
  triggered_by: string | null;
  started_at: string | null;
  finished_at: string | null;
  duration_sec: number | null;
  strict_mode: boolean;
  status: Exclude<PanelRunStatus, null>;
  block_reason: string | null;
  modules: ApiRunModule[];
};

type ApiLatestRunResponse = {
  generated_at: string;
  run: ApiLatestRun;
};

function getApiBaseUrl(): string {
  return "http://127.0.0.1:8000";
}

function getFallbackRun(projectKey: string, project: ApiProjectDetail): PanelLatestRun {
  return {
    run_key: project.latest_run?.run_key ?? `${projectKey}-latest`,
    source: null,
    git_ref: null,
    git_sha: null,
    triggered_by: null,
    started_at: null,
    finished_at: project.latest_run?.finished_at ?? null,
    duration_sec: null,
    strict_mode: false,
    status: project.latest_run?.status ?? "not-run",
    block_reason: project.latest_run?.block_reason ?? null,
  };
}

function mapOverviewProject(project: ApiOverviewProject): PanelProject {
  return {
    project_key: project.project_key,
    project_name: project.project_name,
    repo_path: project.project_key,
    preset: project.preset,
    onboarding_status: project.onboarding_status,
    module_count: project.module_count,
    health: project.health,
    source: null,
    triggered_by: null,
    status: project.check_all,
    block_reason: project.block_reason,
    started_at: null,
    finished_at: project.last_run_at,
    duration_sec: null,
  };
}

function mapProjectDetail(project: ApiProjectDetail, latestRun: ApiLatestRun | null): PanelProjectDetail {
  return {
    project_key: project.project_key,
    project_name: project.project_name,
    repo_path: project.project_key,
    preset: project.preset,
    onboarding_status: project.onboarding_status,
    latest_run: latestRun
      ? {
          run_key: latestRun.run_key,
          source: latestRun.source,
          git_ref: latestRun.git_ref,
          git_sha: latestRun.git_sha,
          triggered_by: latestRun.triggered_by,
          started_at: latestRun.started_at,
          finished_at: latestRun.finished_at,
          duration_sec: latestRun.duration_sec,
          strict_mode: latestRun.strict_mode,
          status: latestRun.status,
          block_reason: latestRun.block_reason,
        }
      : getFallbackRun(project.project_key, project),
    modules: latestRun?.modules ?? project.modules,
  };
}

export async function getPanelProjects(): Promise<PanelProjectsResponse> {
  const response = await fetch(`${getApiBaseUrl()}/api/v1/panel/projects`, {
    cache: "no-store",
  });

  if (!response.ok) {
    throw new Error(`Failed to load panel projects: ${response.status}`);
  }

  const data = (await response.json()) as ApiProjectsResponse;

  return {
    generated_at: data.generated_at,
    projects: data.projects.map(mapOverviewProject),
  };
}

export async function getPanelProject(projectKey: string): Promise<PanelProjectDetailResponse | null> {
  const detailResponse = await fetch(`${getApiBaseUrl()}/api/v1/panel/projects/${projectKey}`, {
    cache: "no-store",
  });

  if (detailResponse.status === 404) {
    return null;
  }

  if (!detailResponse.ok) {
    throw new Error(`Failed to load panel project '${projectKey}': ${detailResponse.status}`);
  }

  const detailData = (await detailResponse.json()) as ApiProjectDetailResponse;

  const latestResponse = await fetch(`${getApiBaseUrl()}/api/v1/panel/projects/${projectKey}/latest`, {
    cache: "no-store",
  });

  let latestRun: ApiLatestRun | null = null;

  if (latestResponse.ok) {
    const latestData = (await latestResponse.json()) as ApiLatestRunResponse;
    latestRun = latestData.run;
  } else if (latestResponse.status !== 404) {
    throw new Error(`Failed to load latest panel run for '${projectKey}': ${latestResponse.status}`);
  }

  return {
    generated_at: detailData.generated_at,
    project: mapProjectDetail(detailData.project, latestRun),
  };
}

File: apps/api/app/schemas/panel.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


PanelCheckerName = Literal["build", "lint", "typecheck", "test", "coverage"]
PanelCheckerStatus = Literal["pass", "fail", "timeout", "skip"]
PanelRunStatus = Literal["pass", "fail", "timeout", "blocked", "not-run"]
PanelOnboardingStatus = Literal["ready", "pending", "blocked", "custom-needed"]
PanelHealthStatus = Literal["healthy", "warning", "failing", "unknown"]


class PanelProjectModule(BaseModel):
    module_name: str = Field(description="Module path, such as apps/api")
    stack: str | None = Field(default=None, description="Detected stack summary")
    language: str | None = Field(default=None, description="Primary language label")


class PanelProjectCommands(BaseModel):
    build: str | None = Field(default=None, description="Build command")
    lint: str | None = Field(default=None, description="Lint command")
    typecheck: str | None = Field(default=None, description="Typecheck command")
    test: str | None = Field(default=None, description="Test command")
    coverage: str | None = Field(default=None, description="Coverage command")


class PanelProjectThresholds(BaseModel):
    coverage_tolerance_pct: float | None = Field(default=None, description="Coverage tolerance percentage")
    coverage_floor_pct: float | None = Field(default=None, description="Coverage floor percentage")
    lint_max_errors: int | None = Field(default=None, description="Maximum lint errors")
    typecheck_max_errors: int | None = Field(default=None, description="Maximum typecheck errors")


class PanelProjectTimeouts(BaseModel):
    build: int | None = Field(default=None, description="Build timeout in seconds")
    test: int | None = Field(default=None, description="Test timeout in seconds")


class PanelProjectNotify(BaseModel):
    feishu_webhook: str | None = Field(default=None, description="Feishu webhook URL")


class PanelProjectRegisterRequest(BaseModel):
    project_key: str = Field(description="Stable project identifier")
    project_name: str = Field(description="Display name")
    repo_path: str = Field(description="Repository path")
    preset: str = Field(description="CLI resolved preset")
    onboarding_status: PanelOnboardingStatus = Field(description="Current onboarding state")
    commands: PanelProjectCommands | None = Field(default=None, description="Project-level commands")
    thresholds: PanelProjectThresholds | None = Field(default=None, description="Threshold configuration")
    timeouts: PanelProjectTimeouts | None = Field(default=None, description="Timeout configuration")
    notify: PanelProjectNotify | None = Field(default=None, description="Notification configuration")
    modules: list[PanelProjectModule] = Field(description="Detected module list")


class PanelProjectRegisterResponse(BaseModel):
    ok: bool = Field(description="Whether the request succeeded")
    project_id: int = Field(description="Resolved project id")
    module_count: int = Field(description="Number of modules upserted")


class PanelBaselineModule(BaseModel):
    module_name: str = Field(description="Module path, such as apps/api")
    baseline_pct: float = Field(description="Effective baseline coverage percentage")


class PanelBaselineUploadRequest(BaseModel):
    project_key: str = Field(description="Stable project identifier")
    updated_at: datetime = Field(description="Baseline update timestamp")
    modules: list[PanelBaselineModule] = Field(description="Baseline rows to upload")


class PanelBaselineUploadResponse(BaseModel):
    ok: bool = Field(description="Whether the request succeeded")
    project_id: int = Field(description="Resolved project id")
    baselines_upserted: int = Field(description="Number of baseline rows upserted")


class PanelCheckerResult(BaseModel):
    checker: PanelCheckerName = Field(description="Checker name")
    status: PanelCheckerStatus = Field(description="Checker execution status")
    detail: str | None = Field(default=None, description="Checker output summary")
    duration_sec: float | None = Field(default=None, description="Checker duration in seconds")


class PanelRunModule(BaseModel):
    module_name: str = Field(description="Module path, such as apps/api")
    stack: str | None = Field(default=None, description="Runtime stack summary")
    language: str | None = Field(default=None, description="Runtime language label")
    status: PanelRunStatus = Field(description="Module status for this run")
    coverage_pct: float | None = Field(default=None, description="Current line coverage")
    baseline_pct: float | None = Field(default=None, description="Baseline echoed in this run")
    coverage_gate_pct: float | None = Field(default=None, description="Coverage gate threshold")
    coverage_delta_pct: float | None = Field(default=None, description="Coverage delta versus baseline")
    coverage_parser: str | None = Field(default=None, description="Coverage parser type")
    block_reason: str | None = Field(default=None, description="Module blocking reason")
    checker_results: list[PanelCheckerResult] = Field(description="Checker result list")


class PanelRunUploadRequest(BaseModel):
    project_key: str = Field(description="Stable project identifier")
    run_key: str = Field(description="Idempotent run identifier")
    source: str | None = Field(default=None, description="Run source")
    git_ref: str | None = Field(default=None, description="Branch or tag")
    git_sha: str | None = Field(default=None, description="Commit SHA")
    triggered_by: str | None = Field(default=None, description="Trigger user or system")
    started_at: datetime | None = Field(default=None, description="Run start timestamp")
    finished_at: datetime | None = Field(default=None, description="Run finish timestamp")
    duration_sec: float | None = Field(default=None, description="Total duration")
    strict_mode: bool = Field(description="Whether strict mode was enabled")
    status: PanelRunStatus = Field(description="Project-level run status")
    block_reason: str | None = Field(default=None, description="Project-level blocking reason")
    modules: list[PanelRunModule] = Field(description="Per-module run snapshots")


class PanelRunUploadResponse(BaseModel):
    ok: bool = Field(description="Whether the request succeeded")
    project_id: int = Field(description="Resolved project id")
    run_id: int = Field(description="Resolved run id")
    modules_upserted: int = Field(description="Number of run modules upserted")
    checker_results_upserted: int = Field(description="Number of checker rows upserted")


class PanelLatestRunRef(BaseModel):
    run_key: str = Field(description="Run identifier")
    status: PanelRunStatus = Field(description="Project-level run status")
    block_reason: str | None = Field(default=None, description="Project-level blocking reason")
    finished_at: datetime = Field(description="Run finish timestamp")


class PanelOverviewProject(BaseModel):
    project_key: str = Field(description="Stable project identifier")
    project_name: str = Field(description="Display name")
    preset: str = Field(description="CLI resolved preset")
    module_count: int = Field(description="Number of modules under the project")
    onboarding_status: PanelOnboardingStatus = Field(description="Current onboarding state")
    health: PanelHealthStatus = Field(description="Derived project health")
    check_all: PanelRunStatus = Field(description="Derived latest project status")
    last_run_at: datetime | None = Field(default=None, description="Latest synchronized run time")
    block_reason: str | None = Field(default=None, description="Latest synchronized block reason")


class PanelProjectOverviewResponse(BaseModel):
    generated_at: datetime = Field(description="Response generation time")
    projects: list[PanelOverviewProject] = Field(description="Overview rows")


class PanelProjectDetail(BaseModel):
    project_key: str = Field(description="Stable project identifier")
    project_name: str = Field(description="Display name")
    preset: str = Field(description="CLI resolved preset")
    onboarding_status: PanelOnboardingStatus = Field(description="Current onboarding state")
    latest_run: PanelLatestRunRef | None = Field(default=None, description="Latest run summary")
    modules: list[PanelRunModule] = Field(description="Latest run module cards")


class PanelProjectDetailResponse(BaseModel):
    generated_at: datetime = Field(description="Response generation time")
    project: PanelProjectDetail = Field(description="Project detail payload")


class PanelLatestRun(BaseModel):
    run_key: str = Field(description="Run identifier")
    source: str | None = Field(default=None, description="Run source")
    git_ref: str | None = Field(default=None, description="Branch or tag")
    git_sha: str | None = Field(default=None, description="Commit SHA")
    triggered_by: str | None = Field(default=None, description="Trigger user or system")
    started_at: datetime | None = Field(default=None, description="Run start timestamp")
    finished_at: datetime | None = Field(default=None, description="Run finish timestamp")
    duration_sec: float | None = Field(default=None, description="Total duration")
    strict_mode: bool = Field(description="Whether strict mode was enabled")
    status: PanelRunStatus = Field(description="Project-level run status")
    block_reason: str | None = Field(default=None, description="Project-level blocking reason")
    modules: list[PanelRunModule] = Field(description="Per-module run snapshots")


class PanelLatestRunResponse(BaseModel):
    generated_at: datetime = Field(description="Response generation time")
    run: PanelLatestRun = Field(description="Latest run payload")
